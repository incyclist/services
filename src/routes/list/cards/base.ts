import { Card   } from "../../../base/cardlist";
import { PromiseObserver } from "../../../base/types/observer";
import { useUserSettings } from "../../../settings";
import { Route } from "../../base/model/route";

export class BaseCard implements Card<Route> {
  

    protected visible=true;
    protected initialized

    setInitialized(val: boolean) {
        this.initialized = val
    }

    reset() {        
        this.setInitialized(false)
        this.setVisible(false)
    }
    

    delete(): PromiseObserver<boolean> {
        throw new Error("Method not implemented.");
    }

    getId(): string {
        throw new Error("Method not implemented.");
    }
    getData(): Route {
        throw new Error("Method not implemented.");
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    setData(_data: Route) {
        throw new Error("Method not implemented.");
    }

    getCardType() {
        throw new Error("Method not implemented.");
    }
    getDisplayProperties() {
        throw new Error("Method not implemented.");
    }

    isVisible(): boolean {
        return this.visible
    }
    setVisible(visible: boolean) {
        this.visible = visible
    }
    equals(card: Card<Route>): boolean {
        if (!card)
            return false
        try {
            return card.getCardType()===this.getCardType() && card.getId()===this.getId()
        }
        catch(err) {
            console.log('~~~~ ERROR',card,err)
            return false
        }

    }

    onMounted(){}

    protected getUserSetting(key:string,defValue?) {
        try {
            const userSettings = useUserSettings();

            const res = userSettings.get(key,defValue)
            return res
        }
        catch (err) {
            return defValue
        }
    }

}


