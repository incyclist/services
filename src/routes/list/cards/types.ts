import { ImportFilter } from "../../../base/cardlist/types";
import { Observer } from "../../../base/types/observer";


export type RouteCardType = 'Import' | 'Route' | 'Free-Ride' | 'ActiveImport';


export const DEFAULT_TITLE = 'Import Route';
export const DEFAULT_FILTERS = [
    { name: 'Routes', extensions: ['gpx', 'epm', 'xml'] },
    { name: 'Tracks', extensions: ['gpx'] },
    { name: 'RLV: ErgoPlanet', extensions: ['epm'] },
    { name: 'RLV: Incyclist, KWT, Rouvy,Virtualtrainer ', extensions: ['xml'] },
];

export interface RouteImportProps {
    title: string;
    filters: Array<ImportFilter>;
    visible: boolean;
}

export interface ActiveImportProps {
    name: string;
    error?: Error
    visible: boolean;
    observer: Observer
}

