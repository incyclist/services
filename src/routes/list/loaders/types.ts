import { Observer } from "../../../base/types/observer";
import { waitNextTick } from "../../../utils";
import { valid } from "../../../utils/valid";
import { RouteApiDescription, RouteApiDetail } from "../../base/api/types";
import { Route } from "../../base/model/route";
import { RouteInfo } from "../../base/types";
import { checkIsLoop, getRouteHash } from "../../base/utils/route";

export interface RouteInfoDBEntry extends RouteInfo {
    pointsEncoded?:string
}

export interface RouteDBApiDescription extends RouteApiDescription {
    type: 'video'|'gpx',
    legacyId?:string,
    originalName?:string
}

interface MinimalDescription {
    country?: string
    title?:string
}

export abstract class Loader<T extends MinimalDescription> { 
    protected loadObserver: Observer

    abstract load():Observer 
    abstract stopLoad():void
    abstract save(route:Route):Promise<void>

    protected abstract buildRouteInfo(descr:T):RouteInfo 

    protected isCompleted(route:Route):boolean {
        const descr = route.description

        if (!valid(descr.points))
            return false;

        if (descr.hasVideo) {
            return   valid(descr.videoUrl) || (descr.requiresDownload && valid(descr.downloadUrl))
        }

        return true;
    }

    protected verifyRouteHash(route:Route) {
        const {description,details} = route

        if (description.points && !description.routeHash) {
            if (details?.routeHash) {
                description.routeHash = details.routeHash
                return;
            }

            const data:RouteApiDetail = details || { id:description.id,title:description.title,points:description.points}
            description.routeHash = getRouteHash( data) 
        }

    }

    protected async verifyCountry(route:Route) {
        const updated = await route.updateCountryFromPoints()
        
        if (updated) {                            
            this.emitRouteUpdate(route)    
        }      
    }

    protected getCountryPrefix(title?:string):string|undefined {
        if (!title)
            return

        if (title.match(/^[A-z]{2}[-_].*/g)) {            
            return title.substring(0,2)
        }
    }

    protected updateRouteCountry( data: RouteInfo, route:{ descr?:T}):void {
        const {descr} = route;

        
        if (descr?.country && !data.country) {
            data.country = descr.country
        }
        if (data.country)
            return;

        const prefix = this.getCountryPrefix(data.title) as string
        if (prefix) {
            data.country = prefix.toLowerCase()
            if (data.category)
                return;
        }       
       
    }

    protected updateRouteTitle( data: RouteInfo, route:{ descr?:T}):void{
        const {descr} = route;
        if (descr && !data.title) {
            data.title = descr.title
        }
        const prefix = this.getCountryPrefix(data.title) as string
        if (prefix) {
            data.title = this.removeCountryPrefix(data.title)
        }

    }

    protected removeCountryPrefix(title?:string):string {
        if (!title)
            return

        if (title.match(/^[A-z][A-z][-_].*/g)) {
            return title.substring(3)
        }
        
    }






    protected emitRouteUpdate( route:Route) {
        if (this.loadObserver)
            this.loadObserver.emit('route.updated',route)
    }
    protected emitRouteAdded( route:Route) {
        if (this.loadObserver)
            this.loadObserver.emit('route.added',route)
    }
    protected emitDone() {
        if (this.loadObserver)
            this.loadObserver.emit('done')
        
        waitNextTick().then(()=>{
            this.loadObserver.reset()
            delete this.loadObserver
        })
    }

    protected addDetails(route:Route, details:RouteApiDetail) {
        route.details = details
        route.description.points = details.points
        
        if (route.description.hasVideo) {
            route.description.requiresDownload = valid(details.downloadUrl)
            route.description.downloadUrl = details.downloadUrl
            if (!valid(route.description.videoUrl)) {
                route.description.videoUrl = details.video.url 
            }
            if (!valid(route.description.videoUrl)) {
                route.description.videoUrl = valid(details.video.file) ? 'video:///'+details.video.file : undefined
            }

            if (!valid(route.description.videoFormat)) {
                route.description.videoFormat = details.video.format
            }
            if (!valid(route.description.previewUrl)) {
                route.description.previewUrl = details.previewUrl                
            }
           
            route.description.hasGpx = details.points.find( p=> p.lat && p.lng)!==undefined
            route.description.next = details.video?.next
            

        }

        if (!valid(route.description.routeHash))
            route.description.routeHash = details.routeHash

        route.description.isLoop = checkIsLoop(route.description.points)

    }


}

export type LoadDetailsTargets = Array<{
    route: Route;
    added: boolean;
}>;

