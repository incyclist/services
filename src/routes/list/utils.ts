import { useUserSettings } from "../../settings"
import { RouteInfo } from "../base/types"
import { RoutesRepoUpdates } from "./types"


const DURATION_2_WEEKS = 1000*60*60*24*14
const DURATION_30_SECONDS = 1000*30

export const updateRepoStats = (ts?:number):void => {
    const settings = useUserSettings()
    const updates = settings.get('repo.routesUpdates',{}) as unknown as RoutesRepoUpdates
    if (updates.current)
        updates.prev = updates.current
    else 
        updates.initial = ts ? ts : Date.now()
    if (ts)
        updates.initial = ts
    updates.current = Date.now()
    settings.set('repo.routesUpdates',updates)
}

export const getRepoUpdates = ():RoutesRepoUpdates => {
    const settings = useUserSettings()
    const updates = settings.get('repo.routesUpdates',{}) as unknown as RoutesRepoUpdates
    return updates
}

export  const checkIsNew = (descr: RouteInfo):boolean => {
    let isNew = false;
    const repoUpdates = getRepoUpdates();

    if (descr.tsImported===repoUpdates.initial)
        return false
    if (repoUpdates.prev) {
        if (!descr.isLocal && !descr.tsLastStart && (descr.tsImported - repoUpdates.initial) > DURATION_30_SECONDS && Date.now() - descr.tsImported < DURATION_2_WEEKS) {
            isNew = true;
        }
    }
    return isNew;
}

