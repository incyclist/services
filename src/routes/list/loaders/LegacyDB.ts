import { JSONObject } from "../../../api"
import { Singleton } from "../../../base/types"
import { Observer, PromiseObserver } from "../../../base/types/observer"
import { valid } from "../../../utils/valid"
import { RouteInfo } from "../../base/types"
import { RouteDBApiDescription } from "./types"
import { DBLoader } from "./DBLoader"

@Singleton
export class RoutesLegacyDbLoader extends DBLoader<RouteDBApiDescription>{
    
    protected loadObserver: Observer
    protected saveObserver: PromiseObserver<void>
    

    constructor () {
        super()
        this.routeDescriptions = []        
    }


    async save(): Promise<void> {
        // not required - routes should only be saved from DBLoader
        return;
    } 

    protected getIsLocal(descr:RouteDBApiDescription):boolean {
        const {type,category} = descr
        if (type==='video') {
            return !valid(category) || category==='Imported'
        }
        if (type==='gpx') {
            return category==='personal'
        }        
    }


    protected buildRouteInfo(descr:RouteDBApiDescription):RouteInfo {
        const { id,legacyId,title,localizedTitle,country,distance,elevation, category,provider, video, points,type,routeHash, originalName} = descr
        
        const data:RouteInfo = { id,legacyId,title,localizedTitle,country,distance,elevation,provider,category,routeHash,originalName}

        data.hasVideo = false;
        data.hasGpx = type==='gpx'
        data.isLocal = this.getIsLocal(descr)
        

        this.updateRouteCountry(data,{descr})
        this.updateRouteTitle(data,{descr})
        // Todo: previewImg (could be generated from video/streetview)

        if (points) data.hasGpx = true;

        if (category?.toLowerCase()==='demo') 
            data.isDemo = true;

        if (video) {
            data.hasVideo = true;
            if (video.format) data.videoFormat = video.format
            if (video.url && !data.videoUrl) data.videoUrl = video.url            
        }

        return data;
    }

    protected buildRouteDBInfo():RouteDBApiDescription {
        throw new Error('not implemented')
    }

    protected isCompleted(): boolean {
        // ensure that loadDetails is not called and 'route.added' events are emitted
        return true        
    }

    protected checkLegacy(descr:RouteDBApiDescription ):RouteDBApiDescription {
        if (descr.routeId) {
            descr.legacyId = descr.id
            descr.id = descr.routeId
            delete descr.routeId
        }
        return descr
    }

    protected async loadDescriptions():Promise<Array<RouteDBApiDescription>> {

        const add = (data:JSONObject, type:'video'|'gpx') => {
            try {
                const routes = data as unknown as Array<RouteDBApiDescription>                
                return routes.map( r=> {
                    r.type = type;
                    return r
                })
            }
            catch(err) {
                return null;
            }

        }

        const videos = this.getVideosRepo().read('routes').then( data=>add(data,'video'))
        const routes = this.getRoutesRepo().read('routes').then( data=>add(data,'gpx'))

        const descriptions:Array<RouteDBApiDescription> = []

        const res = await Promise.allSettled( [videos,routes]) 
        res.forEach( p => {
            if (p.status==='fulfilled' && p.value) {                   
                const data = p.value.map( d=>this.checkLegacy(d))
                
                descriptions.push( ...data)
            }
        })
       
        return descriptions;
    }

    protected async loadDetails():Promise<void> {
        // Legacy DB is only used to load descriptions
        // details should be loaded via DB Loader
        throw new Error('not implemented')
    }

}
