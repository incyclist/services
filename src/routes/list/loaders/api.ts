import { Observer, PromiseObserver } from "../../../base/types/observer";
import IncyclistRoutesApi from "../../base/api";
import { RouteApiDescription, RouteApiDetail } from "../../base/api/types";
import { Route } from "../../base/model/route";
import { RouteInfo } from "../../base/types";
import { RoutesDbLoader } from "./db";
import { Loader } from "./types";
import { LoadDetailsTargets } from "./types";

const valid = (v) => (v!==undefined && v!==null)

export class RoutesApiLoader extends Loader<RouteApiDescription> {

    protected api: IncyclistRoutesApi
    protected loadObserver: Observer
    protected saveObserver: PromiseObserver<void>

    constructor () {
        super()
        this.api = IncyclistRoutesApi.getInstance()        
    }


    load():Observer {
        // no concurrent loads
        if (this.loadObserver)
            return this.loadObserver

        const enrichApi = (v:RouteApiDescription,type)=> ({type, ...v })
        const api = [
            this.api.getRouteDescriptions({type:'gpx'}).then( v=> v.map(e=>enrichApi(e,'gpx'))) ,
            this.api.getRouteDescriptions({type:'video'}).then( v=> v.map(e=>enrichApi(e,'video')))
        ]

        this.loadObserver = new Observer()

        Promise.allSettled( api).then (async res => {

            // join lists
            const items: Array<{route:Route, added:boolean}> = []

            // for loop as we have asyncs here
            for ( let i=0; i<res.length;i++) {
                const p = res[i]

                if (p.status==='fulfilled') {  
                    for (let j=0;j<p.value.length;j++) {
                        const descr = p.value[j]

                    
                        let description;

                        const existing = this.getDescriptionFromDB(descr.routeId||descr.id)
                        const {isDeleted=false} = descr

                        if (existing 
                              && ((existing.version||0)>=(descr.version||0)) 
                              && (!isDeleted || (isDeleted && existing.isDownloaded))
                            ) {

                            const details = await this.getDetailsFromDB(existing.id)
                            if (details) {
                                continue;
                            }
                            description = existing
                            
                        }
                        else {
                            description = this.buildRouteInfo(descr)
                        }

                        if (!description) {
                            continue
                        }


                        
                        const route = new Route( description)

                        const isComplete = this.isCompleted(route)
                        
                        if (isComplete) {
                            this.verifyRouteHash(route)
                            this.emitRouteAdded(route)
                        }

                        if (!existing) {
                            items.push({route,added:isComplete})
                        }
                        else {
                            this.save(route,true)
                        }

                    }
                }
            }

            await this.loadDetails(items)
            
            //delete this.loadObserver
        });

        return this.loadObserver
        
    }

    stopLoad() {
        delete this.loadObserver
    }

    protected getDescriptionFromDB(id:string) {
        const db = new RoutesDbLoader()
        return db.getDescription(id)

    }
    protected async getDetailsFromDB(id:string):Promise<RouteApiDetail> {

        const db = new RoutesDbLoader()        
        return await db.getDetails(id)

    }

    async save(route:Route,enforcedWriteDetails:boolean=false):Promise<void> {
        const db = new RoutesDbLoader()
        await db.save(route,enforcedWriteDetails)
    }


    protected isCompleted(route:Route):boolean {
        const descr = route.description

        if (descr.hasVideo) {
            return valid(descr.points) && ( valid(descr.videoUrl)) || (descr.requiresDownload && valid(descr.downloadUrl))
        }
        else {
            return valid(descr.points)
        }
    }

    protected isPreviewMissing(route:Route) {
        const descr = route.description
        return descr?.hasVideo && !descr?.previewUrl
    }

    protected async loadDetailsFromRepo( items:LoadDetailsTargets):Promise<LoadDetailsTargets>{

        // all missing should be loaded from server
        const promises = items.map ( i => this.getDetailsFromDB (i.route.description.id).then( details => { this.addDetails(i.route,details)}) )

        const res = await Promise.allSettled(promises) 

        const {failed} = this.processLoadDetailsResult(res, items, false);
        return failed
        
    }


    async loadDetails( items:LoadDetailsTargets) {


        // first try to load details from repo
        const failed = await this.loadDetailsFromRepo(items)


        // all missing should be loaded from server
        const promises = failed.map ( i => this.getDetails(i.route) )

        const res = await Promise.allSettled(promises) 

        this.processLoadDetailsResult(res, items, true);

    }

    private processLoadDetailsResult(res: PromiseSettledResult<void>[],items: LoadDetailsTargets, save?:boolean):{success:LoadDetailsTargets,  failed: LoadDetailsTargets} {
        const success:LoadDetailsTargets = []
        const failed: LoadDetailsTargets = []

        res.forEach((pr, idx) => {
            if (pr.status === 'fulfilled') success.push(items[idx]);
            else failed.push(items[idx]);
        });

        success.forEach(item => {
            if (item.added)
                this.emitRouteUpdate(item.route);

            else
                this.emitRouteAdded(item.route);


            this.verifyCountry(item.route);
            this.verifyRouteHash(item.route)

            if (save)
                this.save(item.route, true);

        });

        return {success,failed}
    }

    protected async getDetails(route:Route):Promise<void> {

        const details = await this.api.getRouteDetails(route.description.legacyId||route.description.id)
        this.addDetails(route,details)

            
    }

    protected buildRouteInfo( descr:RouteApiDescription, isLocal?:boolean): RouteInfo {
        const { id,routeId,title,localizedTitle,country,distance,elevation, category,provider, video, points,previewUrl,routeHash,version,isDeleted} = descr
        
        const data:RouteInfo = { title,localizedTitle,country,distance,elevation,provider,category,previewUrl,routeHash,version,isDeleted}

        data.hasVideo = false;
        data.hasGpx = false;
        data.isLocal = isLocal||false
        data.id = routeId || id
        data.legacyId = routeId? id: undefined


        this.updateRouteCountry(data,{descr})
        this.updateRouteTitle(data,{descr})
        // Todo: previewImg (could be generated from video/streetview)

        if (points) data.hasGpx = true;

        if (category?.toLowerCase()==='demo') 
            data.isDemo = true;

        if (video) {
            data.hasVideo = true;
            if (video.format) data.videoFormat = video.format
            if (video.url && !data.videoUrl) data.videoUrl = video.url            
        }

        return data;
    }

 
}