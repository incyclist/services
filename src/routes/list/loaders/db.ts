import { JsonRepository } from "../../../api"
import { Singleton } from "../../../base/types"
import { Observer, PromiseObserver } from "../../../base/types/observer"
import { Route } from "../../base/model/route"
import { RouteInfo } from "../../base/types"
import { RoutesLegacyDbLoader } from "./LegacyDB"
import { RouteInfoDBEntry } from "./types"
import { DBLoader } from "./DBLoader"
import { RouteApiDetail } from "../../base/api/types"
import { waitNextTick } from "../../../utils"
import { addDetails, getTotalElevation, validateDistance, validateSlopes } from "../../base/utils/route"
import { JSONObject } from "../../../utils/xml"
import clone from "../../../utils/clone"

@Singleton
export class RoutesDbLoader extends DBLoader<RouteInfoDBEntry>{
   
    protected loadObserver: Observer
    protected saveObserver: PromiseObserver<void>
    protected videosRepo: JsonRepository
    protected routesRepo: JsonRepository
    protected routeDescriptions: Array<RouteInfoDBEntry>
    protected routeHashes: Record<string,string>
    protected tsLastWrite:number
    protected isDirty:boolean
    protected routesSaveObserver:{ [index:string]:PromiseObserver<void>} = {}
    

    constructor () {
        super()
        this.routeDescriptions = []   
        this.routeHashes = {}
        this.isDirty = true;
    }

    async save(route:Route, enforcedWriteDetails:boolean = false):Promise<void> {
        const stringify = (json) => { try {return JSON.stringify(json)} catch (err) {/* */ console.log('# error',err)} }

        const updatedDescr = this.buildRouteDBInfo(route.description)
        const idx = this.routeDescriptions.findIndex( d=> d.id===route.description.id)

        let changed = false
        if (idx===-1) {
            this.routeDescriptions.push( clone(updatedDescr) )
            this.routeHashes[route.description.id] = stringify(updatedDescr)
            changed = true;
        }
        else { 
            const prev = this.routeHashes[route.description.id]??''
            const updated = stringify(updatedDescr)
            changed = (prev!==updated)
            if (changed) {
                this.routeDescriptions[idx] = clone(updatedDescr)
                this.routeHashes[route.description.id] = updated
            }
            
        }

    
        if (changed) {
            this.isDirty = true;
            this.write()
            this.writeDetails(route)
        }
        else {
            const details = await this.loadDetailRecord(route,false)

            if (!details || enforcedWriteDetails)
                this.writeDetails(route)
        }

    }

    async delete(route:Route):Promise<void> {
        const id = route.description.id
        const idx = this.routeDescriptions.findIndex( d=> d.id===id)
        if (idx==-1)
            throw new Error('route not found')

        this.routeDescriptions.splice(idx,1)

        this.isDirty = true;
        await this.writeRepo()

        await this.deleteRouteDetails(route)

    }

    getDescription (id:string): RouteInfo { 
        const descr = this.routeDescriptions.find( d=>d.id===id)
        return descr
    }



    async getDetails (id:string): Promise<RouteApiDetail> {
        const descr = this.getDescription(id)
        if (!descr)
            return;

        const details = await this.loadDetailRecord(descr)
        return details
    }


    protected async writeRepo() {
        // avoid concurrent updates
        if (this.saveObserver)
            await this.saveObserver.wait()

        const save = async ():Promise<void>=> {
            try {
                await this.routesRepo.write('db',this.routeDescriptions.map(this.buildRouteDBInfo.bind(this)) as JSONObject)
                this.isDirty = false
            }
            catch(err) {
                this.logger.logEvent({message:'could not safe repo',error:err.message })
            }
        }
       
        this.saveObserver = new PromiseObserver( save())
        await this.saveObserver.start()
        process.nextTick( ()=> {delete this.saveObserver})

    }


    protected write() {

        if (this.isDirty && (this.tsLastWrite===undefined || Date.now()-this.tsLastWrite>=1000)) {
            this.isDirty = false;
            this.tsLastWrite = Date.now()            
            this.writeRepo()
        }

        if (this.isDirty && Date.now()-this.tsLastWrite<1000) {
            setTimeout( ()=>{this.write()},  this.tsLastWrite+1000-Date.now())
        }

    }



    protected getLegacyLoader():RoutesLegacyDbLoader {
        return new RoutesLegacyDbLoader()        
    }

    protected buildRouteInfo(descr:RouteInfoDBEntry):RouteInfo {        
        return descr
    }

    protected buildRouteDBInfo(descr:RouteInfo):RouteInfoDBEntry {
        const data = {...descr}
        delete data.points
        return data
    }

    protected async loadFromLegacy():Promise<Array<RouteInfoDBEntry>> {

        return new Promise( done => {
            const observer = this.getLegacyLoader().load();
            const descriptions: Array<RouteInfoDBEntry> = []
            observer.on('route.added',(r)=>{
                
                this.loadDetails(r,false).then( ()=>descriptions.push(r.description))
            })
            observer.on('done',()=>done(descriptions))
        })
    }

    protected async loadDescriptions():Promise<Array<RouteInfoDBEntry>> {
        const stringify = (json) => { try {return JSON.stringify(json)} catch (err) {/* */ console.log('# error',err)} }

        const descriptions = await this.getRoutesRepo().read('db') 
        if (descriptions) {
        
            const routes = descriptions as unknown as Array<RouteInfoDBEntry>
            const cleaned = this.removeDuplicates(routes)

            cleaned.forEach( r=> {
                this.routeHashes[r.id] = stringify(r)
            })
            return cleaned
        }

        const legacy = await this.loadFromLegacy()
        return legacy
    }

    protected removeDuplicates(routes: Array<RouteInfoDBEntry>) {
        const ids = routes.map( r=> r.legacyId || r.id)
        const uniqueIds = ids.filter( (d,pos)=> ids.indexOf(d)===pos)

        const cleaned=[]
        uniqueIds.forEach( id=> {
            const routeWithLegacy = routes.find( r=> r.legacyId===id)
            
            const route = routeWithLegacy || routes.find( r=> r.id===id)
            if (route)
                cleaned.push(route)
        })
        return cleaned
    }


    protected verifyImportDate(routes:Array<RouteInfoDBEntry>)
    {
        
        let changed =false
        routes.forEach( r=> {
            if (!r.tsImported) {
                r.tsImported = Date.now()
                changed = true
            }
        })

        
        if (changed) {
             this.isDirty = true;
             this.write()
        }
        
    }

    protected async loadDetailRecord(target:Route|RouteInfo, log:boolean=true):Promise<RouteApiDetail> {
        
        const description = ((target as Route).description || target) as RouteInfo
        const repo = description?.hasVideo ? this.getVideosRepo() : this.getRoutesRepo()


        const loadFromRepo = async(id, logErrors=true):Promise<RouteApiDetail> => {
            try {

                const res = await repo.read(id) as undefined as RouteApiDetail                
                if (res) {

                    description.originalName = res.title

                    if (description.hasVideo && !description.segments) {
                        description.segments = res.video?.selectableSegments || res.selectableSegments
                    }
        
                }
                else if (logErrors) 
                    this.logger.logEvent({ message: 'could not load route details', id, title: description?.title, reason: 'no data received' })
                return res
            }
            catch (err) {
                this.logger.logEvent({ message: 'could not load route details', id, title: description?.title, reason: err.message, stack: err.stack })
                return null
            }
        }


        let details
        if (description.legacyId) {
            details = await loadFromRepo(description.legacyId,false) 
        }
        if (!details) {
            details = await loadFromRepo(description.id,log) 
        }


        if (!details) {
            return;
        }

        this.validateDetails(details)
        this.validateDescription(description, details)


        return details
    }
    
    private validateDetails(details: any) {
        if (!details.points || !Array.isArray(details.points)) {
            // legacy format?
            details.points = details.decoded
            delete details.decoded
        }

        if (details.localizedTitle && typeof (details.localizedTitle) === 'string') {
            details.localizedTitle = { en: details.localizedTitle }
        }

        validateDistance(details.points)
        validateSlopes(details.points)
    }

    private validateDescription(description: RouteInfo, details: any) {
        if (!description.elevation) {
            description.elevation = getTotalElevation(details)
        }
    }

    protected async loadDetails(route:Route, alreadyAdded?:boolean): Promise<void> {
        const details = await this.loadDetailRecord(route)
        let updated = false

        addDetails(route,details)
        updated ||= this.verifyRouteHash(route)
        updated ||= this.verifyVideoUrl(route)
        
        if (alreadyAdded)
            this.emitRouteUpdate(route)
        this.emitRouteAdded(route)

        this.verifyCountry(route)

        if (updated)
            this.save(route)


    } 

    protected getVideoRepo() {
        if (!this.videosRepo)
            this.videosRepo  =JsonRepository.create('videos')
        return this.videosRepo
    }

    protected getRoutesRepo() {
        if (!this.routesRepo)
            this.routesRepo  =JsonRepository.create('routes')
        return this.routesRepo
    }

    protected getDetailsRepo(route:Route) {
        return route?.description?.hasVideo ? this.getVideoRepo() : this.getRoutesRepo()
    }

    protected async writeDetails(route:Route) {

        const repo = this.getDetailsRepo(route) 
        if (!repo) {
            return;
        }

        const id = route.description.legacyId || route.description.id
        // avoid concurrent updates

        if (this.routesSaveObserver[id]) {
            await this.routesSaveObserver[id].wait()
        }

        const save = async ():Promise<void>=> {
            try {
                await repo.write(id,route.details as undefined as JSONObject)
            }
            catch(err) {
                //console.log(err)
            }
        }
       
        this.routesSaveObserver[id] = new PromiseObserver( save())
        await this.routesSaveObserver[id].start()
        process.nextTick( ()=> {delete this.routesSaveObserver[id]})

    }

    protected async deleteRouteDetails(route:Route) {

        const repo = this.getDetailsRepo(route)
        if (!repo)
            return;

        const id = route.description.id
        if (this.routesSaveObserver[id])
            await this.routesSaveObserver[id].wait()

        const  _deleteRoute = async (id):Promise<void> => {
            await waitNextTick()
    
            try {
                await repo.delete(id)
                waitNextTick().then( ()=> {delete this.routesSaveObserver[id]})
            }
            catch(err) {
                waitNextTick().then( ()=> {delete this.routesSaveObserver[id]})
                throw err
            }    
        }
    

        this.routesSaveObserver[id] = new PromiseObserver( _deleteRoute(id))
        await this.routesSaveObserver[id].start()
    }

}
