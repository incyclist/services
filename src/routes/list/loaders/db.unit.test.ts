import routesData from '../../../../__tests__/data/db/routes.json';
import videosData from '../../../../__tests__/data/db/videos.json';
import sydney from '../../../../__tests__/data/routes/sydney.json'
import holzweiler from '../../../../__tests__/data/rlv/holzweiler.json'
import valdrome from '../../../../__tests__/data/rlv/valdrome.json'

import repoData from '../../../../__tests__/data/db/db.json'
import {  JsonRepository } from '../../../api';
import { Countries } from '../../../i18n/countries';
import { Route } from "../../base/model/route";
import { RoutesDbLoader } from './db';
import { RouteInfo } from '../../base/types';
import { RoutesLegacyDbLoader } from './LegacyDB';
import { JSONObject } from '../../../utils/xml';

class MockRepository extends JsonRepository {
    static create(repoName:string):JsonRepository {
        return super.create(repoName)
    }
    static reset() {
        JsonRepository._instances = {}
        
    }
}


async function dbTest(loader: RoutesDbLoader, routes: Route[],id:string, t:'video'|'gpx' = 'gpx', legacy:boolean=false) {
    const videosRepo = MockRepository.create('videos');
    const routesRepo = MockRepository.create('routes');


    const countries = new Countries();
    countries.getIsoFromLatLng = jest.fn().mockResolvedValue('AU');
    const type = t||'gpx'

    videosRepo['id'] = `${type}-${id}`
    routesRepo['id'] = `${type}-${id}`

    videosRepo.read = jest.fn(async (file) => {
        if (file === 'routes') {
            const route = type==='video' ? videosData.find(r => r.id === id) : undefined;
            return [route] as unknown as JSONObject;
        }
        if (file === id && !legacy)
            return holzweiler as JSONObject;
        if (file === id && legacy)
            return valdrome as unknown as JSONObject

        return null as unknown as JSONObject;

    });

    routesRepo.read = jest.fn(async (file) => {
        if (file === 'db') {
            if (legacy) return null as unknown as JSONObject;

            const route = repoData.find(r => r.id === id);
            return [route] as unknown as JSONObject;
        }


        if (file === 'routes') {            
            const route = type==='gpx' ? routesData.find(r => r.id === id) : undefined;
            return [route] as unknown as JSONObject;
        }
        if (file === id)
            return sydney;

        return null as unknown as JSONObject;
    });

    routesRepo.write = jest.fn();


    
    await new Promise(done => {
        loader.load()
            .on('route.added', route => { 
                routes.push(route); 
            })
            .on('route.updated', done)
            .on('done', () => {                 
                setTimeout(done, 100); 
            });           
    });

    videosRepo.read = jest.fn()
    routesRepo.read = jest.fn()

}






describe('LegacyDBLoader',()=>{
    describe('load',()=>{

  
        let loader:RoutesDbLoader;
        let loaderObj 
        let legacy;
        beforeEach( ()=>{
            loader = loaderObj = new RoutesDbLoader()                    
            legacy = new RoutesLegacyDbLoader()
            
        })

        afterEach( ()=>{
            // cleanup Singletons
            loaderObj.reset()            
            legacy.reset()
            MockRepository.reset()
        })

        test('GPX File from Legacy',async ()=>{
            
            
            const routes:Array<Route> = []
            await dbTest(loader, routes,'5c7a6ae31b1bef04c5854cb3','gpx',true);            

            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.country).toBe('AU')
            expect(descr.isLoop).toBe(true)
            expect(descr.isLocal).toBe(false)
            
            
        })



        test('GPX File from DB',async ()=>{
            const routes:Array<Route> = []

            await dbTest(loader, routes,'5c7a6ae31b1bef04c5854cb3');            
            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.country).toBe('AU')
            expect(descr.isLoop).toBe(true)
            expect(descr.isLocal).toBe(false)
            expect(routes[0]).toMatchSnapshot()

            loaderObj = {...loader}
            expect(loaderObj['loadObserver'] ).toBeUndefined()

        })

        test('Local GPX File from DB',async ()=>{
            const routes:Array<Route> = []

            await dbTest(loader, routes,'9b779bbc-7a20-44f9-b490-76eecac34ee9');            
            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.country).toBe('ES')
            
            expect(descr.isLocal).toBe(true)
            //expect(routes[0]).toMatchSnapshot()            
        })

        test('Local Video File from DB',async ()=>{
            const routes:Array<Route> = []

            const id = '021dd6dd-c08b-431f-a888-2cd3183b1911'
            const before = {...repoData.find( d=>d.id===id)} as RouteInfo

            await dbTest(loader, routes,id,'video');            
            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.points).toBeDefined()

            if (!before.points)
                delete descr.points
            
            expect(routes[0].description).toMatchObject(before)            
        })

        test('Video File from Legacy',async ()=>{
            const routes:Array<Route> = []

            
            const id = 'a07c62c0-61a0-4f81-b55f-e743b9245006'
            const before = {...repoData.find( d=>d.id===id)} as RouteInfo
            delete before.previewUrl

            await dbTest(loader, routes,id,'video',true);            
            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.points).toBeDefined()
            expect(descr.legacyId).toBeDefined()



            if (!before.points)
                delete descr.points
            descr['routeId'] = descr.id
            descr.id = descr.legacyId
            delete descr.legacyId
            
            expect(routes[0].description).toMatchObject({...before,isLoop:true,next:'3aefb91a-be3e-42c1-b7cb-47bc80bebce5'})            
        })


        test('Downloadable Video File from DB',async ()=>{
            const routes:Array<Route> = []

            const id = '8ec98050-9bee-4e76-9a9f-e0c2ab1756f3'

            await dbTest(loader, routes,id,'video');            
            expect(routes.length).toBe(1)
            
            const descr = routes[0].description
            expect(descr.points).toBeDefined()

          
        })


    })

})

