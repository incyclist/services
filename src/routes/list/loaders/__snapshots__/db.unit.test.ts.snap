// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RoutesDbLoader load GPX File from DB 1`] = `
Route {
  "_description": {
    "country": "AU",
    "distance": 3801.452188724582,
    "elevation": 35.19111323845895,
    "hasGpx": true,
    "hasVideo": false,
    "id": "5c7a6ae31b1bef04c5854cb3",
    "isLocal": false,
    "isLoop": true,
    "title": "Sydney Opera House and Botanic Garden",
  },
  "updateCountryFromPoints": [Function],
}
`;
